utils/src/Vulkan/Utils/Reflect.hs

```diff
@@ -0,0 +1,602 @@
+-- maybe can provide a binding map parameter to specify individual binding by name, like `foo [(1, ["inPos", "inColor"]) (2, ["texCoord"])]` speficies that `(binding = 1) inPos, (binding = 1) inColor, (binding = 2) texCoord`
+-- [island pipeline](https://github.com/tgfrerer/island/blob/76d0d38cba74181fa3774cef38aba4d96b6861dc/modules/le_backend_vk/le_pipeline.cpp#L21)
+makePipelineVertexInputStateCreateInfo
+  :: Vector Input -> Maybe (PipelineVertexInputStateCreateInfo '[])
+makePipelineVertexInputStateCreateInfo []     = Nothing
+makePipelineVertexInputStateCreateInfo inputs = Just zero
+  { vertexBindingDescriptions
+  , vertexAttributeDescriptions
+  }
+ where
+  vertexAttributes =
+    join . V.map makeVertexAttribute $ inputs :: Vector VertexAttribute
+  vertexAttributeDescriptions =
+    makeVertexInputAttributeDescriptions vertexAttributes :: Vector
+        VertexInputAttributeDescription
+  vertexBindingDescriptions =
+    makeVertexInputBindingDescriptions vertexAttributes :: Vector
+        VertexInputBindingDescription
+
+makeVertexInputBindingDescriptions
+  :: Vector VertexAttribute -> Vector VertexInputBindingDescription
+makeVertexInputBindingDescriptions =
+  V.fromList
+    . map makeVertexInputBindingDescription
+    . calculate
+    . groupBy ((==) `on` fst)
+    . sortOn fst
+    . extract
+ where
+  extract :: Vector VertexAttribute -> [("binding" ::: Int, "size" ::: Int)]
+  extract =
+    map
+        (   (,)
+        .   fromIntegral
+        .   (binding :: VertexAttribute -> Word32)
+        <*> fromIntegral
+        .   size
+        )
+      . V.toList
+  calculate
+    :: [[("binding" ::: Int, "size" ::: Int)]]
+    -> [("binding" ::: Int, "stride" ::: Int)]
+  calculate = map (liftA2 (,) (fst . head) (sum . (snd <$>)))
+
+makeVertexInputBindingDescription
+  :: ("binding" ::: Int, "stride" ::: Int) -> VertexInputBindingDescription
+makeVertexInputBindingDescription (binding, stride) = zero
+  { binding   = fromIntegral binding
+  , stride    = fromIntegral stride
+  , inputRate = VERTEX_INPUT_RATE_VERTEX
+  }
+
+makeVertexAttribute :: Input -> Vector VertexAttribute
+makeVertexAttribute Input {..} = V.map
+  (\i -> VertexAttribute { binding  = 0
+                         , location = fromIntegral . (+ location) $ i
+                         , size
+                         , format
+                         }
+  )
+  [0 .. count - 1]
+ where
+  count          = maybe 1 V.sum array :: Int
+  (size, format) = convertVertexAttributeType type'
+
+makeVertexInputAttributeDescriptions
+  :: Vector VertexAttribute -> Vector VertexInputAttributeDescription
+makeVertexInputAttributeDescriptions =
+  V.fromList
+    . join
+    . map process
+    . groupBy ((==) `on` (binding :: VertexAttribute -> Word32))
+    . V.toList
+ where
+  process :: [VertexAttribute] -> [VertexInputAttributeDescription]
+  process = snd . foldl
+    (\(nextOffset, acc) cur -> liftA2
```

Comment:
liftA2 makes this code is really hard to follow, sorry :sweat_smile:

Honestly, I still can't assemble it inside my head even after looking around for 15 minutes.

Reply:
the `liftA2` in this place has been removed by using `mapAccumL` to handle the internal state

Reply:
[reader functor blog](https://blog.ssanj.net/posts/2018-03-05-functor-applicative-and-monad-instances-for-reader.html)